Added a general filter function for mean and median filter

angkul07 opened the pull request:
## Description

**What is this PR**
This PR implements a rolling mean filter and a general filter function rolling_filter that can be used for both mean and median filters. The rolling_filter function has a flexible structure, allowing users to specify the method (mean or median). This PR completely remove the [median_filter](https://github.com/neuroinformatics-unit/movement/blob/main/movement/filtering.py) and replace it with the `rolling_filter` function.

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

**What does this PR do?**
- This PR remove the `median_filter` function and replace it with the `rolling_filter`.
- Applies either a rolling mean or rolling median filter to a given `xarray.DataArray`.


## References

This PR is related to the issue #454 and pull request #429

## How has this PR been tested?

This PR has been tested locally and pass all test cases. This function implemented in this PR has also been tested on PR #429  and generate the similar result as generated in the PR.

## Is this a breaking change?
No

## Does this PR require an update to the documentation?
Yes, this PR require updates to the documentation. Documentation for various examples and API references need to be change. But documentation is not updated yet.

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

sfmig commented:
Looking good!

Re names @niksirbi, what about `moving_filter`? 

I find "moving" a bit more clear than "rolling" but that may be personal. I have seen people use "running" too.

niksirbi commented:
> I find "moving" a bit more clear than "rolling" but that may be personal. I have seen people use "running" too.

I do like "moving", because people tend to use "moving average" a lot. That said, "rolling" has the advantage of sticking to the `xarray` terminology (and matches the underlying func we are using).

angkul07 commented:
@niksirbi you're right. Its a good opportunity to add other filters in the `filtering.py`. I will go ahead with your implementation and make the relevant changes in the `test_filtering.py` and update the smoothing example. 

@sfmig I was initially confused about the naming of the function. So, I use the rolling_filter because 'rolling' was mentioned in the #454. Should I change the function name?

@niksirbi I have one question, if I update the smoothing example, then should I update the relevant documentation of smoothing example?

niksirbi commented:
> @sfmig I was initially confused about the naming of the function. So, I use the rolling_filter because 'rolling' was mentioned in the #454. Should I change the function name?

Which name do you find most intuitive @angkul07? Curious to hear your opinion as an outsider on this.

> @niksirbi I have one question, if I update the smoothing example, then should I update the relevant documentation of smoothing example?

What do you mean by "documentation of the smoothing example"?  Do you mean the text that appears between the code blocks? I think it will need some updating, yes. We should add a note mentioning that even though we use this function to apply a rolling median filter, one could also use it for other statistics such as the mean, etc., via the `statistic` parameter. Did I understand your question correctly?

angkul07 commented:
z
> Which name do you find most intuitive @angkul07? Curious to hear your opinion as an outsider on this.
I think since we decided to have many filters, I think aggregate_filter() or stat_filter() are most intuitive. 


> What do you mean by "documentation of the smoothing example"? Do you mean the text that appears between the code blocks? I think it will need some updating, yes. We should add a note mentioning that even though we use this function to apply a rolling median filter, one could also use it for other statistics such as the mean, etc., via the `statistic` parameter. Did I understand your question correctly?
no no, I was asking about the documentation of this [Smooth pose tracks](https://movement.neuroinformatics.dev/examples/smooth.html#sphx-glr-examples-smooth-py) example given on the website. Since this example use the median filter. So, I think it will be a nice idea to change its documentation, with new filters. I was asking about it.

niksirbi commented:
> I think since we decided to have many filters, I think aggregate_filter() or stat_filter() are most intuitive.

Thanks for sharing your thoughts on this. These are all viable options, alongside the ones mentioned above, but I don't particularly love any of them. To avoid getting stuck in endless back-and-forth discussions about naming, I'd say to stick with the current `rolling_filter` name for now, unless someone comes up with a truly convincing alternative until we merge this PR.

> no no, I was asking about the documentation of this [Smooth pose tracks](https://movement.neuroinformatics.dev/examples/smooth.html#sphx-glr-examples-smooth-py) example given on the website. Since this example use the median filter. So, I think it will be a nice idea to change its documentation, with new filters. I was asking about it.

We are referring to the same thing. I think eventually yes, it would be nice to also showcase other options in this example, like mean vs median vs max, and observe how they affect the data. But I think that we should hold off these changes for another PR, because we've already broadened the scope of this one. In this PR, it's enough to make sure that the "Smooth pose tracks" example uses the updated syntax for the `rolling_filter` function, and that it includes a mention of the other `statistic` options. I mean adding a line that says something like this:

```md
# Here we use the default ``statistic="median"`` option, which is a sensible
# choice for smoothing time-series data while being robust to outliers.
# You can also use the ``rolling_filter`` function to compute the rolling
# mean, maximum, and minimum values (instead of the median), by 
# setting ``statistic`` to equal ``"mean"``, ``"max"``, or ``"min"``,
# respectively.
```

angkul07 commented:
> Thanks for sharing your thoughts on this. These are all viable options, alongside the ones mentioned above, but I don't particularly love any of them. To avoid getting stuck in endless back-and-forth discussions about naming, I'd say to stick with the current `rolling_filter` name for now, unless someone comes up with a truly convincing alternative until we merge this PR.

Yah, that's make sense.



> We are referring to the same thing. I think eventually yes, it would be nice to also showcase other options in this example, like mean vs median vs max, and observe how they affect the data. But I think that we should hold off these changes for another PR, because we've already broadened the scope of this one. In this PR, it's enough to make sure that the "Smooth pose tracks" example uses the updated syntax for the `rolling_filter` function, and that it includes a mention of the other `statistic` options. I mean adding a line that says something like this:

Thanks for clearly my doubt. I think this makes sense, in other orgs too, some PRs are created for documentation update. Should I open another PR related to documentation?

I have made the relevant changes in `test_filtering.py` and smoothing example. You can review it.

niksirbi commented on examples/smooth.py:
```diff
@@ -110,48 +110,55 @@ def plot_raw_and_smooth_timeseries_and_psd(
-# Using the :func:`movement.filtering.median_filter` function on the
-# ``position`` data variable, we can apply a rolling window median filter
+# Using the :func:`movement.filtering.rolling_filter` function on the
+# ``position`` data variable, we can apply a rolling window filter
```
```suggestion
# ``position`` data variable, we can apply a rolling median filter
```

niksirbi commented on examples/smooth.py:
```diff
@@ -110,48 +110,55 @@ def plot_raw_and_smooth_timeseries_and_psd(
 # %%
 # We see from the printed report that the dataset has no missing values
 # neither before nor after smoothing. Let's visualise the effects of the
-# median filter in the time and frequency domains.
+# ``median`` statistic in the time and frequency domains.
```
```suggestion
# neither before nor after smoothing. Let's visualise the effects of
# applying the rolling median filter in the time and frequency domains.
```

niksirbi commented on examples/smooth.py:
```diff
@@ -110,48 +110,55 @@ def plot_raw_and_smooth_timeseries_and_psd(
 
 # %%
-# We see that the median filter has removed the "spikes" present around the
+# We see that the median statistic has removed the "spikes" present around the
 # 14 second mark in the raw data. However, it has not dealt the big shift
```
```suggestion
# We see that applying the filter has removed the "spikes" present around the
# 14 second mark in the raw data. However, it has not dealt with the big shift
```

niksirbi commented on examples/smooth.py:
```diff
@@ -110,48 +110,55 @@ def plot_raw_and_smooth_timeseries_and_psd(
 # affecting the low frequency components.
 #
-# This illustrates what the median filter is good at: removing brief "spikes"
+# This illustrates what median statistic is good at: removing brief "spikes"
```
```suggestion
# This shows what the rolling median is good for: removing brief "spikes"
```

niksirbi commented on examples/smooth.py:
```diff
@@ -110,48 +110,55 @@ def plot_raw_and_smooth_timeseries_and_psd(
-# We can control the behaviour of the median filter
-# via two parameters: ``window`` and ``min_periods``.
+# We can control the behaviour of the rolling filter
+# via two parameters: ``window``, ``min_periods`` and ``statistic`` which was
```
```suggestion
# via three parameters: ``window``, ``min_periods`` and ``statistic`` which was
```

niksirbi commented on examples/smooth.py:
```diff
@@ -163,20 +170,20 @@ def plot_raw_and_smooth_timeseries_and_psd(
 # see that there are some missing values, indicated as "nan" in the
 # printed dataset.
-# Let's apply the median filter over a 0.1-second window (3 frames)
+# Let's apply the median statistic over a 0.1-second window (3 frames)
```
```suggestion
# Let's apply the rolling median filter over a 0.1-second window (3 frames)
```

angkul07 commented:
@niksirbi in my recent commit, I get the following error by pre-commit:

`examples/mouse_eye_movements.py:17: error: Module "movement.filtering" has no attribute "median_filter"  [attr-defined]
Found 1 error in 1 file (checked 84 source files)`

It seems like that this mouse eye movement example was added recently and it uses the median_filter. So, should I update mouse eye movement example with rolling_filter and commit it just like smoothing example?

niksirbi commented on examples/smooth.py:
```diff
@@ -187,16 +194,16 @@ def plot_raw_and_smooth_timeseries_and_psd(
 # %%
 # We see that this time the number of NaN values has decreased
 # across all keypoints.
-# Let's visualise the effects of the median filter in the time and frequency
+# Let's visualise the effects of the median statistic in the time and frequency
```
```suggestion
# across all keypoints. Let's visualise
# the effects of the rolling median filter in the time and frequency
```

niksirbi reviewed (changes requested):
Hey @angkul07, I've highlighted some further suggestion.

One of the problems is that there are many leftover mentions of the now obsolete `method` in the example, which need to be replaced by `statistic`. It seems that you've realised that yourself and are in the process of fixing it.

The other suggestions have to do with the choice of words in the example.
Sorry for being so pedantic here, but we're very particular with the language we use in the examples, and we strive for consistent use of language across the projects. 

> It seems like that this mouse eye movement example was added recently and it uses the median_filter. So, should I update mouse eye movement example with rolling_filter and commit it just like smoothing example?

You are right about that, and actually it's not only this new example, but many other changes have happened across the project which may lead to conflicts with the changes you've made here. These "merge conflicts" are quite common in actively developed projects and are often a pain to deal with. I'd say in this situation it's best to just to implement the suggestions I've given in my review + update the median filter in the eye movement example to use the new syntax. After that I will take over the PR and  deal with any remaining merge conflicts.

angkul07 commented:
> One of the problems is that there are many leftover mentions of the now obsolete `method` in the example, which need to be replaced by `statistic`. It seems that you've realised that yourself and are in the process of fixing it.

Hello @niksirbi, I didn't realise that there were still some mentions were left. I made the mistake in hurry. Sorry about that.

> The other suggestions have to do with the choice of words in the example.
Sorry for being so pedantic here, but we're very particular with the language we use in the examples, and we strive for consistent use of language across the projects.

I am new to the org so I think it will take me some time to this. Thanks for mentioning it, I will keep it in mind.

As you suggested, I updated the mouse eye example to use the new syntax and implemented the changes in smoothing example you suggested. You can review it now. Thanks.

angkul07 commented:
> These "merge conflicts" are quite common in actively developed projects and are often a pain to deal with.

@niksirbi, I just realised that you are write about these merge conflicts. When I started working on the updates, I was up-to date with the main branch and pre-commit was giving me no error. But after I finished the changes and push my commits, the movement repo was update and now pre-commit is giving me new errors. Its very hard to keep up with this fast development :sweat_smile:.

niksirbi commented:
Hey @angkul07, sorry for the delay on this. I will now take over and deal with the merge conflicts myself here, because resolving them requires holding in your head lots of different PRs that were merged recently, and making some executive decisions. That task is even hard for me.

All your contributions and commits will be preserved in this process. Please don't push any more changes to this branch while I undertake this "surgery".

And thanks for all your work, I think this feature will be definitely useful.

angkul07 commented:
Hi @niksirbi, thanks for guiding me throughout this process. I understand the hurdles of maintaining a open source project. All the nui packages are in the begining stage so its more difficult and niu have a small team. So, huge salute to whole niu team for maintaining these packages.

It was fun working with you and I am happy to contribute in the movement. I hope we can work again during GSOC :smiley:. Thanks once again.

niksirbi reviewed (approved):
This is good to merge now!

I dealt with the merge conflicts and also added a deprecation warning with an informative message, in case people try to use the old `median_filter` syntax.
